src/lib/coda_base/transaction_logic.ml

```diff
@@ -611,6 +622,103 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
             ; (source_location, source_account) ]
           , `Source_timing source_timing
           , Undo.User_command_undo.Body.Payment {previous_empty_accounts} )
+      | Create_new_token _ ->
+          (* NOTE: source and receiver are definitionally equal here. *)
+          let fee_payer_account =
+            try
+              let balance =
+                Option.value_exn
+                  (Balance.sub_amount fee_payer_account.balance
+                     (Amount.of_fee constraint_constants.account_creation_fee))
+              in
+              let fee_payer_account = {fee_payer_account with balance} in
+              let timing =
+                Or_error.ok_exn
+                  (validate_timing ~txn_amount:Amount.zero
+                     ~txn_global_slot:current_global_slot
+                     ~account:fee_payer_account)
+              in
+              {fee_payer_account with timing}
+            with exn -> raise (Reject (Error.of_exn exn))
+          in
+          let%map receiver_location, receiver_account =
+            get_with_location ledger receiver
+          in
+          if not (receiver_location = `New) then
+            failwith
+              "Token owner account for newly created token already exists?!?!" ;
+          let receiver_account = {receiver_account with token_owner= true} in
+          ( [ (fee_payer_location, fee_payer_account)
+            ; (receiver_location, receiver_account) ]
+          , `Source_timing receiver_account.timing
+          , Undo.User_command_undo.Body.Create_new_token
+              {created_token= next_available_token} )
+      | Create_token_account _ ->
+          let fee_payer_account =
+            try
```

Comment:
the `try .. with` here appears to be the same as in the `Create_new_token` clause. Can that be factored out?